File: 0436_find_right_interval/find_right.java
class Solution {
    // Time: O(n * log(n))
    // Space: O(n)
    public int[] findRightInterval(int[][] intervals) {
        TreeMap<Integer, Integer> starts = new TreeMap<>();
        int res[] = new int[intervals.length];
        for (int i = 0; i < intervals.length; i++) {
            starts.put(intervals[i][0], i);
        }
        for (int i = 0; i < intervals.length; i++) {
            Map.Entry<Integer, Integer> pos = starts.ceilingEntry(intervals[i][1]);
            res[i] = pos == null ? -1 : pos.getValue();
        }
        return res;
    }
}

File: 0436_find_right_interval/find_right.py
from typing import List


def find_right_interval(intervals: List[List[int]]) -> List[int]:
    """
    Time: O(n * log(n))
    Space: O(n)
    """
    l = sorted((start, i) for i, (start, _) in enumerate(intervals))
    res = []
    for _, end in intervals:
        r = bisect.bisect_left(l, (end,))
        res.append(l[r][1] if r < len(l) else -1)
    return res
